doc/adr/0002-steppable-cek-machine.md

```diff
@@ -0,0 +1,235 @@
+# ADR 2: Steppable CEK machine
+
+Date: 2022-10
+
+## Authors
+
+Marty Stumpf <marty.stumpf@iohk.io>  
+Ziyang Liu <ziyang.liu@iohk.io>
+
+## Status
+
+Proposed
+
+## Context
+
+In order to have a minimal viable product of a debugger for Plutus, we need a CEK machine that will give us more information for debugging than our current one.
+
+Implementing the steppable CEK machine is a non-trivial task and involves some design decisions:
+
+- should the debugging machine be a separate one?
+- should we modify the existing CEK machine to support stepping, or implement something entirely new?
+- how do we ensure performance and conformance?
+
+## Decision: A separate machine
+
+We came to the conclusion that the debugging machine should be a separate one. The main reason is that we don't want to slow down the production machine.
+
+## Argument: A separate machine
+
+The debugging machine will give the most accurate information if it was the one in production. However, the drawback of that is that we have to care about its performance.
+
+We only care about performance if the steppable CEK machine will *replace* the production machine instead of being a separate one.
+
+Our current CEK machine's performance is sensitive to changes. For example, exporting the `computeCek` function from the module causes the CEK machine to become slower by up to 25%. Refactoring the machine in preparation for the stepper also shows significant slow downs. See [PR#4909](https://github.com/input-output-hk/plutus/pull/4909). We have enough evidence to show that replacing the machine without slow downs is impossible.
+
+We consciously chose to keep up the production CEK machine's performance by adding a separate machine for debugging. We will mitigate the drawback of this by keeping the implementation as close to the production machine as possible.
+
+## Alternative: polymorphic compute/return steps
+
+This approach has been suggested to potentially not introduce slow downs while not having a separate debugging CEK machine.
+
+As long as the debugging machine has the same type, we can alter `computeCek`/`returnCek` to be polymorphic over a type-level `Bool` specifying if we’re in debug mode or not. Then we demote it to the term level in the definition of `computeCek`/`returnCek` and branch on the `Bool` thus implementing different logic depending on whether we're in debug mode or not. This promotion to the type level allows us to statically instantiate the `Bool` in an instance and thus make GHC compile the whole worker of the CEK machine twice: once in debug mode and once in production mode. Theoretically, GHC will take care to remove all the dead debug code when in production mode.
```

Comment:
I don't think this approach actually works, since a steppable `computeCek`/`returnCek` won't have the same types as the normal ones. And I don't imagine the debug code to be a superset of the production code, which is what this approach assumes (hence using a Bool flag to turn debug code off in production).

Or maybe I misunderstood it; in either case, if you describe an approach here it is helpful to have some code snippet.

Reply:
See [this](https://github.com/input-output-hk/plutus/pull/4909#issuecomment-1286189092) comment. If you agree it seems to make sense, we can include its content in the proposal (or just a link to it).

Reply:
(it's only one, probably simplest, way of doing it. We could as well use `Either (Proxy a) (Proxy b)` instead of `Bool` and sneak into that argument any type information we want. Or have more arguments to the type class that only the `Bool` one)

Reply:
I think we should try it, of course.